In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
df = pd.read_csv('adult.csv')
df.head()

39          State-gov   77516   Bachelors   13        Never-married  \
0  50   Self-emp-not-inc   83311   Bachelors   13   Married-civ-spouse   
1  38            Private  215646     HS-grad    9             Divorced   
2  53            Private  234721        11th    7   Married-civ-spouse   
3  28            Private  338409   Bachelors   13   Married-civ-spouse   
4  37            Private  284582     Masters   14   Married-civ-spouse   

         Adm-clerical   Not-in-family   White     Male   2174   0   40  \
0     Exec-managerial         Husband   White     Male      0   0   13   
1   Handlers-cleaners   Not-in-family   White     Male      0   0   40   
2   Handlers-cleaners         Husband   Black     Male      0   0   40   
3      Prof-specialty            Wife   Black   Female      0   0   40   
4     Exec-managerial            Wife   White   Female      0   0   40   

    United-States   <=50K  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3            Cuba   <=50K  
4   United-States   <=50K

In [4]:
df.shape

(32560, 15)

In [5]:
df.columns = ['Age','Workclass','Fnlwgt','Education','Education_Num','Martial_Status','Occupation','Relationship',
              'Race','Sex','Capital_Gain','Capital_Loss','Hours_Per_Week','Native_Country','Income']
df.head()

Age          Workclass  Fnlwgt   Education  Education_Num  \
0   50   Self-emp-not-inc   83311   Bachelors             13   
1   38            Private  215646     HS-grad              9   
2   53            Private  234721        11th              7   
3   28            Private  338409   Bachelors             13   
4   37            Private  284582     Masters             14   

        Martial_Status          Occupation    Relationship    Race      Sex  \
0   Married-civ-spouse     Exec-managerial         Husband   White     Male   
1             Divorced   Handlers-cleaners   Not-in-family   White     Male   
2   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
3   Married-civ-spouse      Prof-specialty            Wife   Black   Female   
4   Married-civ-spouse     Exec-managerial            Wife   White   Female   

   Capital_Gain  Capital_Loss  Hours_Per_Week  Native_Country  Income  
0             0             0              13   United-States   <=50K  
1             0             0              40   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40            Cuba   <=50K  
4             0             0              40   United-States   <=50K

In [6]:
df.dtypes

Age                int64
Workclass         object
Fnlwgt             int64
Education         object
Education_Num      int64
Martial_Status    object
Occupation        object
Relationship      object
Race              object
Sex               object
Capital_Gain       int64
Capital_Loss       int64
Hours_Per_Week     int64
Native_Country    object
Income            object
dtype: object

In [7]:
obj_cols = [i for i in df.columns if df[i].dtypes=='object']
obj_cols

['Workclass',
 'Education',
 'Martial_Status',
 'Occupation',
 'Relationship',
 'Race',
 'Sex',
 'Native_Country',
 'Income']

In [8]:
from sklearn.preprocessing import LabelEncoder

In [9]:
lb = LabelEncoder()
for i in obj_cols:
    df[i] = lb.fit_transform(df[i])

In [10]:
df.dtypes

Age               int64
Workclass         int64
Fnlwgt            int64
Education         int64
Education_Num     int64
Martial_Status    int64
Occupation        int64
Relationship      int64
Race              int64
Sex               int64
Capital_Gain      int64
Capital_Loss      int64
Hours_Per_Week    int64
Native_Country    int64
Income            int64
dtype: object

In [11]:
df.isnull().sum()

Age               0
Workclass         0
Fnlwgt            0
Education         0
Education_Num     0
Martial_Status    0
Occupation        0
Relationship      0
Race              0
Sex               0
Capital_Gain      0
Capital_Loss      0
Hours_Per_Week    0
Native_Country    0
Income            0
dtype: int64

In [12]:
x = df.iloc[:,:-1]
y = df.iloc[:,-1]
print(type(x))
print(type(y))
print(x.shape)
print(y.shape)

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>
(32560, 14)
(32560,)


In [13]:
x.head()

Age  Workclass  Fnlwgt  Education  Education_Num  Martial_Status  \
0   50          6   83311          9             13               2   
1   38          4  215646         11              9               0   
2   53          4  234721          1              7               2   
3   28          4  338409          9             13               2   
4   37          4  284582         12             14               2   

   Occupation  Relationship  Race  Sex  Capital_Gain  Capital_Loss  \
0           4             0     4    1             0             0   
1           6             1     4    1             0             0   
2           6             0     2    1             0             0   
3          10             5     2    0             0             0   
4           4             5     4    0             0             0   

   Hours_Per_Week  Native_Country  
0              13              39  
1              40              39  
2              40              39  
3              40               5  
4              40              39

In [14]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: Income, dtype: int64

In [15]:
from sklearn.model_selection import train_test_split

In [49]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.20)    
print(x_train.shape)                                                       
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(26048, 14)
(6512, 14)
(26048,)
(6512,)


## Model Building

In [50]:
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score

In [51]:
def test_metrics(ytest,ypred):
    cm = confusion_matrix(ytest,ypred)
    print('Accuracy Score = ',accuracy_score(ytest,ypred))
    print('\nConfusion Matrix = \n',cm)
    print('\nClassification Report = \n',classification_report(ytest,ypred,zero_division=0))
    print('Misclassification Percentage = ',(1-accuracy_score(ytest,ypred))*100)
    
def tt_score(model):
    print('Training Score = ', model.score(x_train,y_train))
    print('Testing Score = ', model.score(x_test,y_test))

### a. Decision Tree Classifier

In [52]:
from sklearn.tree import DecisionTreeClassifier

In [71]:
m1 = DecisionTreeClassifier(criterion='entropy',max_depth=9,min_samples_split=12)
m1.fit(x_train,y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=9, min_samples_split=12)

In [72]:
tt_score(m1)

Training Score =  0.8616400491400491
Testing Score =  0.855497542997543


In [73]:
ypred_m1 = m1.predict(x_test)
print(ypred_m1)

[1 0 0 ... 0 0 1]


In [74]:
test_metrics(y_test,ypred_m1)

Accuracy Score =  0.855497542997543

Confusion Matrix = 
 [[4565  405]
 [ 536 1006]]

Classification Report = 
               precision    recall  f1-score   support

           0       0.89      0.92      0.91      4970
           1       0.71      0.65      0.68      1542

    accuracy                           0.86      6512
   macro avg       0.80      0.79      0.79      6512
weighted avg       0.85      0.86      0.85      6512

Misclassification Percentage =  14.4502457002457


### b. Random Forest Classifier

In [75]:
from sklearn.ensemble import RandomForestClassifier

In [98]:
m2 = RandomForestClassifier(n_estimators=70,criterion='gini',max_depth=9)
m2.fit(x_train,y_train)

RandomForestClassifier(max_depth=9, n_estimators=70)

In [99]:
tt_score(m2)

Training Score =  0.8633292383292384
Testing Score =  0.8596437346437347


In [100]:
ypred_m2 = m2.predict(x_test)
print(ypred_m2)

[1 0 0 ... 0 0 1]


In [101]:
test_metrics(y_test,ypred_m2)

Accuracy Score =  0.8596437346437347

Confusion Matrix = 
 [[4771  199]
 [ 715  827]]

Classification Report = 
               precision    recall  f1-score   support

           0       0.87      0.96      0.91      4970
           1       0.81      0.54      0.64      1542

    accuracy                           0.86      6512
   macro avg       0.84      0.75      0.78      6512
weighted avg       0.85      0.86      0.85      6512

Misclassification Percentage =  14.035626535626534


### c. Logistic Regression

In [102]:
from sklearn.linear_model import LogisticRegression

In [103]:
m3 = LogisticRegression(max_iter=1000)
m3.fit(x_train,y_train)

LogisticRegression(max_iter=1000)

In [104]:
tt_score(m3)

Training Score =  0.8028255528255528
Testing Score =  0.8022113022113022


In [105]:
ypred_m3 = m3.predict(x_test)
print(ypred_m3)

[1 0 0 ... 0 0 1]


In [106]:
test_metrics(y_test,ypred_m3)

Accuracy Score =  0.8022113022113022

Confusion Matrix = 
 [[4651  319]
 [ 969  573]]

Classification Report = 
               precision    recall  f1-score   support

           0       0.83      0.94      0.88      4970
           1       0.64      0.37      0.47      1542

    accuracy                           0.80      6512
   macro avg       0.73      0.65      0.67      6512
weighted avg       0.78      0.80      0.78      6512

Misclassification Percentage =  19.77886977886978


### d. KNN Classifier

In [107]:
from sklearn.neighbors import KNeighborsClassifier

In [108]:
m4 = KNeighborsClassifier(n_neighbors=19)
m4.fit(x_train,y_train)

KNeighborsClassifier(n_neighbors=19)

In [109]:
tt_score(m4)

Training Score =  0.8039004914004914
Testing Score =  0.8063574938574939


In [110]:
ypred_m4 = m4.predict(x_test)
print(ypred_m4)

[1 0 0 ... 0 0 1]


In [111]:
test_metrics(y_test,ypred_m4)

Accuracy Score =  0.8063574938574939

Confusion Matrix = 
 [[4886   84]
 [1177  365]]

Classification Report = 
               precision    recall  f1-score   support

           0       0.81      0.98      0.89      4970
           1       0.81      0.24      0.37      1542

    accuracy                           0.81      6512
   macro avg       0.81      0.61      0.63      6512
weighted avg       0.81      0.81      0.76      6512

Misclassification Percentage =  19.36425061425061


### e. SVM Classifier

In [ ]:
from sklearn.svm import SVC

In [ ]:
m5 = SVC(kernel='linear',C=10)
m5.fit(x_train,y_train)

In [ ]:
tt_score(m5)

In [ ]:
ypred_m5 = m5.predict(x_test)
print(ypred_m5)

In [ ]:
test_metrics(y_test,ypred_m5)

# <font color='red'>Conclusion</font>

### <u>Random Forest Classifier</u> is the best performing model in terms of accuracy (~86%).<br>